src/transport_performance/osm/validate_osm.py

```diff
@@ -0,0 +1,682 @@
+* Count all features in a pbf file by type (node, way, relation or area)
+* Return IDs of the above feature types
+* Extract tags for the above features
+* Find coordinates for node or way features
+* Plot the coordinates of a given list of node or way IDs
+"""
+import osmium
+from pathlib import Path
+from typing import Union
+import pandas as pd
+import geopandas as gpd
+from shapely import Point
+import folium
+
+from transport_performance.utils.defence import (
+    _check_item_in_iter,
+    _check_iterable,
+    _type_defence,
+    _is_expected_filetype,
+)
+
+# ---------utilities-----------
+
+
+def _compile_tags(osmium_feature):
+    """Return tag name value pairs.
+
+    Parameters
+    ----------
+    osmium_feature : osmium.osm.types
+        Either a node, way, relation or area osmium type.
+
+    Returns
+    -------
+    tagdict: dict
+        A dictionary of tag name & values.
+
+    """
+    return {tag.k: tag.v for tag in osmium_feature.tags}
+
+
+def _filter_target_dict_with_list(
+    targets: dict, _list: list, search_key: str, accepted_keys: list
+) -> dict:
+    """Select a target dictionary and filter it with a list of keys.
+
+    Parameters
+    ----------
+    targets : dict
+        A dictionary of feature_type: attribute dicts.
+    _list : list
+        A list of keys, OSM IDs in this context. Integer keys are expected.
+    search_key : str
+        The key value to search with.
+    accepted_keys : list
+        Valid key values specific to the context of the method calling this
+        function.
+
+    Returns
+    -------
+    filtered_dict.
+        Dictionary with keys filtered to IDs available in `_list`.
+
+    Raises
+    ------
+    ValueError
+        `search_key` is not in `accepted_keys`.
+        `filtered_dict` is empty. No tags for the combination of `_list` and
+        `search_key` were found.
+    TypeError
+        `_list` is not a list.
+        `accepted_keys` is not a list.
+        Elements of `_list` are not integer.
+        Elements of `accepted_keys` are not string.
+        `search_key` is not a string.
+    KeyError
+        `search_key` is not present in `targets.keys()`
+
+    """
+    _type_defence(targets, "targets", dict)
+    _check_iterable(_list, "_list", list, exp_type=int)
+    _type_defence(search_key, "search_key", str)
+    _check_iterable(accepted_keys, "accepted_keys", list, exp_type=str)
+    _check_item_in_iter(search_key, accepted_keys, "search_key")
+
+    feat = search_key.lower().strip()
+    try:
+        targ_dict = targets[feat]
+    except KeyError:
+        raise KeyError(
+            f"`search_key`: {feat} did not match keys in "
+            f"`targets`: {targets.keys()}"
+        )
+
+    filtered_dict = dict(
+        (id, targ_dict[id]) for id in _list if id in targ_dict
+    )
+    if len(filtered_dict) == 0:
+        raise ValueError(
+            "No tags found. Did you specify the correct search_key?"
+        )
+
+    return filtered_dict
```

Comment:
Agreed. On reviewing the purposes I felt this could be tidied up a lot. New approach ditches the accepted_keys parameter altogether and implements a utility method for FindLocations that will check the user is asking for anything other than node or way in the methods and raise a NotImplementedError.